pkg/ingester/incremental_transfer.go

```diff
@@ -0,0 +1,260 @@
+	"github.com/weaveworks/common/user"
+	"golang.org/x/net/context"
+)
+
+var (
+	blockedRanges = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_ingester_blocked_ranges",
+		Help: "The current number of ranges that will not accept writes by this ingester.",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(blockedRanges)
+}
+
+// TransferChunksSubset accepts chunks from a client and moves them into the local Ingester.
+func (i *Ingester) TransferChunksSubset(stream client.Ingester_TransferChunksSubsetServer) error {
+	i.userStatesMtx.Lock()
+	defer i.userStatesMtx.Unlock()
+
+	fromIngesterID, seriesReceived, err := i.acceptChunksFromStream(acceptChunksOptions{
+		States: i.userStates,
+		Stream: stream,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := stream.SendAndClose(&client.TransferChunksResponse{}); err != nil {
+		level.Error(util.Logger).Log("msg", "Error closing SendChunks stream", "from_ingester", fromIngesterID, "err", err)
+		return err
+	}
+
+	// Target ingester may not have had any streams to send.
+	if seriesReceived == 0 {
+		return nil
+	} else if fromIngesterID == "" {
+		level.Error(util.Logger).Log("msg", "received TransferChunks request with no ID from ingester")
+		return fmt.Errorf("no ingester id")
+	}
+
+	return nil
+}
+
+// GetChunksSubset accepts a get request from a client and sends all chunks from the serving ingester
+// that fall within the given range to the client.
+func (i *Ingester) GetChunksSubset(req *client.GetChunksRequest, stream client.Ingester_GetChunksSubsetServer) error {
+	userStatesCopy := i.userStates.cp()
+	if len(userStatesCopy) == 0 {
+		level.Info(util.Logger).Log("msg", "nothing to transfer")
+		return nil
+	}
+
+	// Block the ranges locally so we don't receive traffic for it anymore. The caller should decide
+	// when ranges should be unblocked: we should continue to reject writes for as long as we may
+	// receive them. When the joining token has been completely inserted into the ring, it will
+	// be safe to remove the blocks.
+	i.BlockRanges(req.Ranges)
+
+	return i.pushChunksToStream(pushChunksOptions{
+		States:        userStatesCopy,
+		Stream:        stream,
+		MarkAsFlushed: req.Move,
+
+		IncludeFunc: func(pair fingerprintSeriesPair) bool {
+			return inAnyRange(pair.series.token, req.Ranges)
+		},
+	})
+}
+
+// BlockRanges configures a range of token values to be blocked. When a range
+// is blocked, the Ingester will no longer accept pushes for any streams whose
+// token falls within the blocked ranges. Unblocking the range re-enables those
+// pushes.
+//
+// Blocked ranges will automatically be unblocked after the RangeBlockPeriod
+// configuration variable. This acts as a failsafe to prevent blocked ranges
+// sitting around forever if a joining ingester crashes, as writes will continue
+// to go to us and get rejected for as long as the blocked range exists.
+func (i *Ingester) BlockRanges(ranges []ring.TokenRange) {
```

Comment:
One thing I'm worried about is transfers taking longer than `cfg.RangeBlockPeriod`. If it does, the the range would become unblocked due to this expiration and the ingester would become available for writes again.

What would happen to the new ingester receiving tokens? I think if it immediately retried, that would defeat the purpose of the `cfg.RangeBlockPeriod` as `BlockRanges` would be called again.

Reply:
From @rfratto:

> 1. if a join fails, you'll still join the ring, but you won't have as much data as you should. it'll cause some spillover and desynced chunks.
> 2. if a leave fails, you flush to GCS. also causes some desynced chunks.
> 3. if a transfer is slow (or if a transferring node crashes), you risk more writes being rejected because the blocked range gets released later. we really need to set this value to the longest period we're comfortable tolerating rejections. the tradeoff here is the risk between rejecting data and having spillover.

These alleviated my concerns here.